tailwarden/komiser
# issue 454, add support for AWS AutoScalingGroups

~~**DO NOT MERGE, THIS REQUIRES FURTHER REVIEW AND AGREEMENT**~~

## Problem

This solves the feature request https://github.com/tailwarden/komiser/issues/454

## Solution

Implements the functionality to discover AutoScalingGroups

## Changes Made

- Added a struct that discovers ASGs.
- It requires a struct that satisfies the newly created AutoScalingGroup interface. This allows unit-testing it if not all (I don't think it's possible to achieve more with the current design) at least the meat of the functionality
- I also added a mock of such an interface using vanilla mockery 

## How to Test

From what I see in the mod file, the application has been written in Go 1.18, while I currently work with 1.20.
All CI operations should be easily reproducible and should not depend on local versions.

To execute the test, please run:

```bash
docker run --rm \
	--volume $(PWD):/app \
	--workdir /app \
	golang:1.18-alpine \
	ash -c "go mod tidy && CGO_ENABLED=0 go test ./providers/... -cover"
```

I would extend the same practice to `vet`, `fmt`, and all the tests that need to be executed before the build process.

The tests, which only cover one file, are successful.

```bash
ok      github.com/tailwarden/komiser/providers/aws/ec2 0.008s  coverage: 8.4% of statements
```

I also see many lines reporting other errors which I don't know where they come from.
Is this a known issue?

```bash
 github.com/tailwarden/komiser/providers/aws/cloudfront
providers/aws/cloudfront/distributions.go:54:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/cloudfront/distributions.go:80:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/cloudfront/distributions.go:106:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/lambda
providers/aws/lambda/functions.go:49:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/lambda/functions.go:74:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/apigateway
providers/aws/apigateway/apis.go:57:4: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/sns
providers/aws/sns/topics.go:79:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/sqs
providers/aws/sqs/queues.go:54:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/sqs/queues.go:80:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
providers/aws/sqs/queues.go:106:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/aws/s3
providers/aws/s3/buckets.go:56:4: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/digitalocean/k8s
providers/digitalocean/k8s/kubernetes.go:49:16: fmt.Sprintf format %d has arg kubernetesCluster.ID of wrong type string
# github.com/tailwarden/komiser/providers/digitalocean/network
providers/digitalocean/network/vpcs.go:19:3: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
# github.com/tailwarden/komiser/providers/digitalocean/kubernetes
providers/digitalocean/kubernetes/kubernetes.go:44:16: fmt.Sprintf format %d has arg kubernetesCluster.ID of wrong type string
providers/digitalocean/kubernetes/kubernetes.go:49:16: fmt.Sprintf format %d has arg kubernetesCluster.ID of wrong type string
# github.com/tailwarden/komiser/providers/linode/compute
providers/linode/compute/instances.go:47:16: fmt.Sprintf format %s has arg instance.ID of wrong type int
providers/linode/compute/lke_clusters.go:47:16: fmt.Sprintf format %s has arg lkeCluster.ID of wrong type int
# github.com/tailwarden/komiser/providers/linode/storage
providers/linode/storage/databases.go:30:16: fmt.Sprintf format %s has arg database.ID of wrong type int
providers/linode/storage/volumes.go:47:16: fmt.Sprintf format %s has arg volume.ID of wrong type int
# github.com/tailwarden/komiser/providers/linode/networking
providers/linode/networking/firewalls.go:47:16: fmt.Sprintf format %s has arg firewall.ID of wrong type int
providers/linode/networking/node_balancers.go:47:16: fmt.Sprintf format %s has arg nodeBalancer.ID of wrong type int
# github.com/tailwarden/komiser/providers/aws
providers/aws/aws.go:74:5: github.com/sirupsen/logrus.Warn call has possible formatting directive %s
ok      github.com/tailwarden/komiser/providers/aws/ec2 0.008s  coverage: 8.4% of statements
```

## Notes

Pending

## Checklist

- [ ] Pending

## Reviewers

@[username of the reviewer]
